### PR TITLE
[actors] Stickiness actor pattern

### PR DESCRIPTION
**Pull request by EDsCODE**
## Changes

*Please describe.*  
- apply actor pattern to stickiness insight
*If this affects the frontend, include screenshots.*  

## How did you test this code?

*Please describe.*
- add tests for groups and move to api test

### CODE REVIEW

File: posthog/types.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 from posthog.models.filters.filter import Filter
 from posthog.models.filters.path_filter import PathFilter
 from posthog.models.filters.retention_filter import RetentionFilter
+from posthog.models.filters.session_recordings_filter import SessionRecordingsFilter
 from posthog.models.filters.stickiness_filter import StickinessFilter
 
-FilterType = Union[Filter, PathFilter, RetentionFilter, StickinessFilter]
+FilterType = Union[Filter, PathFilter, RetentionFilter, StickinessFilter, SessionRecordingsFilter]
```
Comment (macobo) [inline posthog/types.py:9]: This shouldn't be in this PR right?

This "implicitly" makes column_optimizer support session recordings filters which isn't the case and is scary as heck.

File: ee/clickhouse/queries/groups_join_query.py [hunk 2]
```diff
@@ -4,20 +4,21 @@
...
 
-    _filter: Union[Filter, PathFilter, RetentionFilter]
+    _filter: FilterType
     _team_id: int
     _column_optimizer: ColumnOptimizer
 
     def __init__(
         self,
-        filter: Union[Filter, PathFilter, RetentionFilter],
+        filter: FilterType,
```
Comment (macobo) [inline ee/clickhouse/queries/groups_join_query.py:21]: How does this work given that column_optimizer does not accept FilterType?

File: ee/clickhouse/queries/stickiness/stickiness_actors.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    @cached_property
+    def is_aggregating_by_groups(self) -> bool:
+        return self.entity.math == "unique_group"
+
+    def actor_query(self) -> Tuple[str, Dict]:
+        events_query, event_params = StickinessEventsQuery(
+            entity=self.entity, filter=self.filter, team_id=self._team.pk
+        ).get_query()
+        query = f"""
```
Comment (macobo) [inline ee/clickhouse/queries/stickiness/stickiness_actors.py:30]: Nit: Just inline this in the return, code would look better

File: ee/clickhouse/queries/stickiness/stickiness_actors.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        SELECT DISTINCT aggregation_target FROM ({events_query}) WHERE num_intervals = %(stickiness_day)s
+        {'LIMIT %(limit)s' if self.filter.limit else ''}
```
Comment (macobo) [inline ee/clickhouse/queries/stickiness/stickiness_actors.py:32]: Nit: null checks should use `is not None` as in python other values are also falsy

File: ee/clickhouse/queries/session_recordings/clickhouse_session_recording_list.py [hunk 5]
```diff
@@ -131,6 +132,28 @@ class ClickhouseSessionRecordingList(ClickhouseEventQuery):
     LIMIT %(limit)s OFFSET %(offset)s
     """
 
+    def __init__(
```
Comment (macobo) [inline ee/clickhouse/queries/session_recordings/clickhouse_session_recording_list.py:135]: Q: Why is this change needed? This is unrelated to stickiness and creates a lot of noise in the code.

File: ee/clickhouse/queries/retention/retention_event_query.py [hunk 6]
```diff
@@ -21,10 +22,29 @@ class RetentionEventsQuery(ClickhouseEventQuery):
     _event_query_type: RetentionQueryType
     _trunc_func: str
 
-    def __init__(self, event_query_type: RetentionQueryType, *args, **kwargs):
+    def __init__(
```
Comment (macobo) [inline ee/clickhouse/queries/retention/retention_event_query.py:25]: Q: Why is this change needed? This is unrelated to stickiness and creates a lot of noise in the code.

**Review by macobo — COMMENTED** [review]
Some of the __init__ code/filters code is suspect to me - adds noise, bad copied defaults and other bs.

File: posthog/types.py (hunk 1 above)
Comment (macobo) [inline posthog/types.py:9]: This is not needed - `SessionRecordingsFilter` is already a `Filter`

**Review by macobo — COMMENTED** [review]
Some of the __init__ code/filters code is suspect to me - adds noise, bad copied defaults and other bs.

**Comment by EDsCODE** [comment]
Was getting trigger happy with wanting to unify init and filter typing. Changes weren't good and inappropriate for this PR. Removed them

File: ee/clickhouse/queries/stickiness/stickiness_actors.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+from posthog.models.team import Team
+
+
+class ClickhouseStickinessActors(ActorBaseQuery):
+    entity: Entity
+    _filter: StickinessFilter
+
+    def __init__(self, team: Team, entity: Optional[Entity], filter: StickinessFilter):
+        if not entity:
+            raise ValueError("Entity is required")
```
Comment (macobo) [inline ee/clickhouse/queries/stickiness/stickiness_actors.py:18]: Should this be a validationError or is a 500 warranted?

**Review by macobo — APPROVED** [review]

File: ee/clickhouse/queries/stickiness/stickiness_actors.py (hunk 7 above)
Comment (EDsCODE) [reply ee/clickhouse/queries/stickiness/stickiness_actors.py]: just removed this check. the init shouldn't take an optional
